### PR TITLE
Resolves # 25

### LINKED ISSUE #25: Create a dev docker-compose (useful for development)

### PR DESCRIPTION
**Pull request by akashvarun10**
This pull request adds Docker support for the Vidur application with a Dockerfile and Docker Compose setup. It configures the application and PostgreSQL database, ensuring automated builds and data persistence. Use `docker compose up --build` to start the services.

Resolves #25

### CODE REVIEW

**Comment by akashvarun10** [comment]
I will be working on adding for production version of docker, once the app is ready..

**Comment by amandesai01** [comment]
We don't want to send .env file in docker, we can just use docker-compose to set the env variables

**Comment by akashvarun10** [comment]
If we use `localhost` reference in the `docker-compose.yml` is incorrect for `inter-container communication`. Within Docker, containers communicate using service names defined in Docker Compose. Using db instead of localhost ensures that the app container correctly resolves and connects to the db container.

And since this setup is only for development and not production, this approach should be fine for local testing and development purposes I guess

**Comment by akashvarun10** [comment]
Using environment variables in this manner for a development setup allows for easy launching and testing of the application and since this setup is only for development and not production, this approach should be fine for local testing and dev purposes

**Comment by amandesai01** [comment]
Yes I understand that, point being, we don't want to have 2 ways to setup `env` since it might be confusing. It's always better to rely on single source of truth.

**Comment by akashvarun10** [comment]
Okay, I have modified it . Can you check it now, I just changed it create env from env.example only

**Comment by amandesai01** [comment]
Code should not be copied over, directory should be mounted (so that code changes still work)
Something like this:
```yml
volumes:
      - .:/usr/src/app
      - /usr/src/app/node_modules
```

**Comment by amandesai01** [comment]
Also, we don't want to hardcode env vars like you did in database. I ran this in detail through chatgpt, and found this:
```yml
   env_file:
      - .env
    environment:
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_DB: ${POSTGRES_DB}
```

**Comment by amandesai01** [comment]
Can you do something like this?

**Comment by akashvarun10** [comment]
Sure, I'll change it

**Comment by amandesai01** [comment]
Also, I am pretty curious about something and we would need to fix this (hopefully you can research on this):
Size of postgres image is 400mb
Size of Vidur codebase including node modules is 200mb
But, size of docker image: 2.4GB!!??

**Comment by amandesai01** [comment]
Also, please claim the issue and link PR

**Comment by akashvarun10** [comment]
> Also, I am pretty curious about something and we would need to fix this (hopefully you can research on this): Size of postgres image is 400mb Size of Vidur codebase including node modules is 200mb But, size of docker image: 2.4GB!!??

I'll check it out, it mostly for the layers of the containers, unused large files and node_modules i guess there is `.dockerignore`, I'll see what i can do

**Comment by akashvarun10** [comment]
> Also, please claim the issue and link PR

sure

**Comment by akashvarun10** [comment]
Also, is it fine if i change this to `NUXT_DB_HOST="db"`? Because in Docker providing it as localhost leads to internal communication error

**Comment by amandesai01** [comment]
Just request a review when you are ready.

**Comment by akashvarun10** [comment]
It's Done actually

**Comment by amandesai01** [comment]
We will also need to mount ./tmp dir since it will be used to store blobs in dev mode. I will do it.

**Comment by akashvarun10** [comment]
Sure, I'll try check how can we reduce the container size..

**Comment by akashvarun10** [comment]
Here I can submit a review but can't request it  I think you should add yourself on reviewer because i am not getting any option

**Comment by akashvarun10** [comment]
mounted temp storage for blob

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+COPY package.json yarn.lock ./
+
+RUN yarn install
+
+COPY . .
+
+EXPOSE 3001
+
+CMD ["sh", "-c", "yarn migration:apply && yarn dev"]
+
```
Comment (amandesai01) [inline Dockerfile:14]:
```suggestion
FROM node:18.20-alpine

WORKDIR /app

COPY package.json yarn.lock ./

RUN yarn install
```

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    depends_on:
+      - db
+    volumes:
+      - .:/app
+      - /app/node_modules
+      - ./tmp:/app/tmp  # Mounting the ./tmp directory for blob storage
+    command: sh -c "cp .env.example .env && yarn migration:apply && yarn dev"
+
+volumes:
+  db_data:
```
Comment (amandesai01) [inline docker-compose.yaml:38]:
```suggestion
version: '3.8'

services:
  db:
    image: postgres:15.1-alpine
    container_name: database
    env_file:
      - .env
    environment:
      POSTGRES_USER: ${NUXT_DB_USER:-postgres}
      POSTGRES_PORT: ${NUXT_DB_PORT:-5432}
      POSTGRES_PASSWORD: ${NUXT_DB_PASSWORD:-postgrespw}
      POSTGRES_DB: ${NUXT_DB_DATABASE:-vidur}
    ports:
      - "5432:5432"
    volumes:
      - vidur_data:/var/lib/postgresql/data

  app:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: vidur_app
    env_file:
      - .env
    environment:
      NUXT_DB_HOST: db
      NUXT_DB_PORT: ${NUXT_DB_PORT:-5432}
      NUXT_DB_USER: ${NUXT_DB_USER:-postgres}
      NUXT_DB_PASSWORD: ${NUXT_DB_PASSWORD:-postgrespw}
      NUXT_DB_DATABASE: ${NUXT_DB_DATABASE:-vidur}
    ports:
      - "3001:3001"
    depends_on:
      - db
    volumes:
      - .:/app
      - ./src/public/tmp:/app/src/public/tmp  # Mounting the ./src/public/tmp directory for blob storage
    command: sh -c "yarn migration:apply && yarn dev"

volumes:
  vidur_data:
```

File: docs/README.md [hunk 3]
```diff
@@ -48,6 +48,11 @@ yarn migration:apply  # Create database schema.
 yarn dev              # Finally, start the dev server.
 ```
 
+Or Simply by Using Docker and Navigate to vidur then you can do as follows:
```
Comment (amandesai01) [inline docs/README.md:51]:
```suggestion
Or simply by using Docker Compose.
```

File: docs/README.md [hunk 4]
```diff
@@ -48,6 +48,11 @@ yarn migration:apply  # Create database schema.
...
+```sh
+docker compose up --build
```
Comment (amandesai01) [inline docs/README.md:53]:
```suggestion
docker compose up
```

**Review by amandesai01 — CHANGES_REQUESTED** [review]

**Pull request merged**